src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.http.sink.batch;
+
+import com.google.common.base.Strings;
+import com.google.gson.internal.bind.JsonTreeWriter;
+import com.google.gson.stream.JsonWriter;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.format.StructuredRecordStringConverter;
+import org.json4s.jackson.Json;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * MessageBuffer is used to store the structured records in a buffer till the batch size is reached.
+ * Once the batch size is reached, the records are converted to the appropriate format and appended to the message.
+ * The message is then returned to the HTTPRecordWriter.
+ */
+public class MessageBuffer {
+  private static final String REGEX_HASHED_VAR = "#s*(\\w+)";
+  private final List<StructuredRecord> buffer;
+  private final String jsonBatchKey;
+  private final Boolean shouldWriteJsonAsArray;
+  private final String delimiterForMessages;
+  private final String charset;
+  private final String customMessageBody;
+  private final Function<List<StructuredRecord>, String> messageFormatter;
+  private final String contentType;
+
+  public MessageBuffer(
+          String messageFormat, String jsonBatchKey, boolean shouldWriteJsonAsArray,
+          String delimiterForMessages, String charset, String customMessageBody
+  ) {
+    this.jsonBatchKey = jsonBatchKey;
+    this.delimiterForMessages = delimiterForMessages;
+    this.charset = charset;
+    this.shouldWriteJsonAsArray = shouldWriteJsonAsArray;
+    this.customMessageBody = customMessageBody;
+    this.buffer = new ArrayList<>();
+    switch (messageFormat.toLowerCase()) {
```

Comment:
Might be valuable to make `messageFormat` an enum instead.

Reply:
Added MessageFormatType